aten/src/ATen/native/cuda/Normalization.cu

```diff
@@ -0,0 +1,444 @@
+    invstd = save_invstd[plane];
+  } else {
+    mean = static_cast<accscalar_t>(running_mean[plane]);
+    invstd = static_cast<accscalar_t>(1) / std::sqrt(static_cast<accscalar_t>(running_var[plane]) + epsilon);
+  }
+
+  accscalar_t weight_val = weight.size(0) > 0 ? static_cast<accscalar_t>(weight[plane]) : accscalar_t(1);
+  accscalar_t norm = accscalar_t(1) / N;
+
+  // Compute two values across (batch, x/y/z) in one pass:
+  // 1. Sum(grad_output)
+  // 2. DotProduct(input - mean, grad_output)
+  GradOp<scalar_t, accscalar_t> g(mean, input, grad_output);
+  Float2<scalar_t, accscalar_t> res = reduce<Float2<scalar_t, accscalar_t>, GradOp<scalar_t, accscalar_t>, PackedTensorAccessor<scalar_t, 3, at::RestrictPtrTraits>>(g, grad_output, plane);
+  accscalar_t grad_output_sum = res.v1;
+  accscalar_t dot_p = res.v2;
+
+  accscalar_t grad_mean = grad_output_sum * norm;
+  accscalar_t proj_scale = dot_p * norm * invstd * invstd;
+  accscalar_t grad_scale = invstd * weight_val;
+
+  if (grad_input.data() != NULL) {
+    for (int batch = 0; batch < grad_output.size(0); ++batch) {
+      for (int x = threadIdx.x; x < grad_output.size(2); x += blockDim.x) {
+        scalar_t go = grad_output[batch][plane][x];
+        if (train) {
+          scalar_t inp = input[batch][plane][x];
+          accscalar_t proj = (inp - mean) * proj_scale;
+          grad_input[batch][plane][x] = static_cast<scalar_t>((go - proj - grad_mean) * grad_scale);
+        } else {
+          grad_input[batch][plane][x] = static_cast<scalar_t>(go * grad_scale);
+        }
+      }
+    }
+  }
+
+  if (grad_weight.size(0) > 0) {
+    if (threadIdx.x == 0) {
+      grad_weight[plane] = static_cast<scalar_t>(dot_p * invstd);
+    }
+  }
+
+  if (grad_bias.size(0) > 0) {
+    if (threadIdx.x == 0) {
+      grad_bias[plane] = static_cast<scalar_t>(grad_output_sum);
+    }
+  }
+}
+
+// TensorAccessor in which the last dimensions are collapsed or expanded as needed
+template <typename scalar_t, int64_t dim>
+static PackedTensorAccessor<scalar_t, dim, at::RestrictPtrTraits> reshaped_packed_accessor(const Tensor& t) {
+  // undefined...
+  if (! t.defined()) {
+    const std::vector<int64_t> zeros(dim);
+    return PackedTensorAccessor<scalar_t, dim, at::RestrictPtrTraits>(nullptr, zeros.data(), zeros.data());
+  }
+  int64_t in_dim = t.dim();
+  if (in_dim == dim) {
+    return t.packed_accessor<scalar_t, dim, at::RestrictPtrTraits>();
+  }
+
+  AT_CHECK(in_dim < dim || t.is_contiguous(), "need contiguous or <= 3d tensor");
+  std::vector<int64_t> sizes(dim);
+  std::vector<int64_t> strides(dim);
+  for (int i = 0; i < in_dim || i < dim; ++i) {
+    if (i < dim && i < in_dim) {
+      sizes[i] = t.size(i);
+      strides[i] = t.stride(i);
+    } else if (i < dim) {
+      sizes[i] = 1;
+      strides[i] = 0;
+    } else {
+      sizes[dim - 1] *= t.size(i);
+      strides[dim -1] = 1;
+    }
+  }
+  // evil trick to get adjusted 2d tensors to have large dimension last
```

Comment:
So, IIUC, this is where you pick the larger of dim0 and dim2 to reduce over, right?

Reply:
Indeed.